### PR TITLE
Pluggable driver

### PR DESCRIPTION
**Pull request by nathan-jenan-rancher**
This change gives rancher the ability to accept pluggable kontainer engine drivers by
adding the necessary field information to the drivers and adding support for building
the driver options over grpc.
    
Issue:
https://github.com/rancher/rancher/issues/12698

Depends on:
https://github.com/rancher/types/pull/413

### CODE REVIEW

File: cluster/cluster.go [hunk 1]
```diff
@@ -282,8 +290,10 @@ func (c *Cluster) restore() error {
 }
 
 // NewCluster create a cluster interface to do operations
-func NewCluster(driverName, addr, name string, configGetter ConfigGetter, persistStore PersistentStore) (*Cluster, error) {
-	rpcClient, err := types.NewClient(driverName, addr)
+func NewCluster(driverName, name string, listenAddr string, configGetter ConfigGetter, persistStore PersistentStore) (*Cluster, error) {
+	var err error
```
Comment (StrongMonkey) [inline cluster/cluster.go]: what's the point of declaring this err?

File: types/rpc_server.go [hunk 2]
```diff
@@ -100,20 +97,42 @@ func (s *GrpcServer) GetCapabilities(ctx context.Context, in *Empty) (*Capabilit
...
+		errChan <- err
 	}
 	return
 }
+
+// ServeOrDie serves a grpc server or kills the process
+func (s *GrpcServer) ServeOrDie(listenAddr string) {
+	listen, err := net.Listen("tcp", listenAddr)
+	if err != nil {
+		logrus.Fatalf("%v", err)
```
Comment (StrongMonkey) [inline types/rpc_server.go]: Nit: this is equal to logrus.Fatal(err)

**Comment by StrongMonkey** [comment]
LGTM

File: service/service.go [hunk 3]
```diff
@@ -241,10 +289,160 @@ func (e *engineService) Update(ctx context.Context, name string, clusterSpec v3.
...
+			return "", fmt.Errorf("no driver for name: %v", r.Name)
+		}
+
+		addr := make(chan string)
+		errChan := make(chan error)
+		r.Server = types.NewServer(driver, addr)
+		go r.Server.Serve(listenAddress, errChan)
+
+		// if the error hasn't appeared after 5 seconds assume it won't error
+		time.Sleep(5 * time.Second)
```
Comment (cjellick) [inline service/service.go]: switch this to a time channel

Comment (nathan-jenan-rancher) [reply service/service.go]: Done.

**Comment by cjellick** [comment]
LGTM

**Pull request merged**